compiler/src/iree/compiler/Codegen/Common/SpecializeExports.cpp

```diff
@@ -0,0 +1,463 @@
+          builder.setInsertionPoint(ordinalOp);
+          Value assumedOperand =
+              builder
+                  .create<IREE::Util::AssumeIntOp>(loc, ordinalOp.getOperand(),
+                                                   builder.getArrayAttr(range))
+                  .getResult(0);
+          ordinalOp.setOperand(assumedOperand);
+        }
+      }
+
+      builder.create<IREE::HAL::ReturnOp>(loc, exportCondition);
+    }
+    // Current function is still the original function, just with a new symbol
+    // name.
+    currentExport = newExport;
+  }
+
+  return;
+}
+
+/// Walks the function |func| exported by |exportOp| and looks for a tilable
+/// operation annotated with specialization ranges. The range annotation can
+/// be accesses with |helper|.
+static void specializeExportedFunctionByRangeAttribute(
+    IREE::HAL::ExecutableExportOp exportOp, func::FuncOp func,
+    RangeAttrHelper helper, llvm::SmallDenseSet<int64_t> &ordinals) {
+  TilingInterface specializationRoot;
+  IREE::Util::MultiIntAssumptionArrayAttr specializationRanges;
+
+  // Walk for the first op specifying specialization ranges. Its unclear what
+  // to do with multiple annotations (which one takes precedence etc.) so for
+  // now take the first.
+  func.walk([&](TilingInterface op) {
+    if (auto maybeRanges = helper.getAttr(op)) {
+      specializationRanges = maybeRanges;
+      specializationRoot = op;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+
+  if (!specializationRoot) {
+    return;
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Specialization root:\n\t";
+    llvm::dbgs() << specializationRoot << "\n";
+  });
+
+  // Remove the attribute now before it potentially gets cloned across multiple
+  // function copies.
+  helper.removeAttr(specializationRoot);
+
+  specializeExportedFunction(exportOp, func, specializationRoot,
+                             specializationRanges, ordinals);
+}
+
+namespace {
+struct SpecializeExportsPass final
+    : impl::SpecializeExportsPassBase<SpecializeExportsPass> {
+public:
+  using Base::Base;
+
+  void runOnOperation() override {
+    IREE::HAL::ExecutableVariantOp variant = getOperation();
+
+    auto *codegenDialect =
+        getContext().getLoadedDialect<IREE::Codegen::IREECodegenDialect>();
+    RangeAttrHelper helper =
+        codegenDialect->getSpecializationRangesAttrHelper();
+
+    ModuleOp innerModule = variant.getInnerModule();
+    if (!innerModule) {
+      // Nothing to do if function definitions aren't provided.
+      return;
+    }
+
+    SmallVector<IREE::HAL::ExecutableExportOp, 1> exports(
+        variant.getExportOps());
+
+    // Get the set of exported ordinals. We need to know which ordinals are
+    // available when introducing new exports.
+    llvm::SmallDenseSet<int64_t> ordinalSet;
+    for (auto exportOp : exports) {
+      IntegerAttr ordinalAttr = exportOp.getOrdinalAttr();
+      if (!ordinalAttr) {
+        return;
```

Comment:
I don't think you want to silently return here - if ordinals are required for the pass you should exportOp.emitError and return signalPassFailure

Reply:
I'm unsure of the exact meaning of a missing ordinal, but failure to specialize shouldn't be a compiler failure IMO. I changed it to walk the variant and drop all annotations given a missing ordinal.

Better would be a warning. Do we have a preferred way to emit warnings? I see `op.emitWarning` in a few places but don't want them to be unsilenceable.

Reply:
I think a compiler error is appropriate - if the rule is that ordinals need to be assigned by the time this function runs then it's an error to run the pass without them assigned. If the rule is that ordinals are optional then you want to handle everything you can (not bail early). If the rule is that ordinals are never assigned then the pass does nothing :P

We shouldn't end up with some ordinals set and some not on the same executable. We could add a verifier check for that.